{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cyvcf2\n",
    "import requests\n",
    "import csv\n",
    "import pickle\n",
    "\n",
    "def parse_vcf_entry(entry):\n",
    "    \"\"\"\n",
    "    Parses a single VCF entry to extract essential information for each alternate allele.\n",
    "    \"\"\"\n",
    "    base_data = {\n",
    "        \"chrom\": entry.CHROM,\n",
    "        \"pos\": entry.POS,\n",
    "        \"ref\": entry.REF,\n",
    "        \"depth\": entry.INFO.get('DP'),  # Extracting depth from VCF entry\n",
    "    }\n",
    "    return [{\"alt\": alt, **base_data} for alt in entry.ALT]\n",
    "\n",
    "def query_ensembl(variant):\n",
    "    \"\"\"\n",
    "    Queries the Ensembl VEP endpoint using SPDI notation for each variant.\n",
    "    \"\"\"\n",
    "    url = \"http://grch37.rest.ensembl.org/vep/human/region\"\n",
    "    headers = {\"Content-Type\": \"application/json\"}\n",
    "    adjusted_position = variant['pos'] + 1\n",
    "    spdi_notation = f\"{variant['chrom']}:{adjusted_position}:{variant['ref']}:{variant['alt']}\"\n",
    "\n",
    "    payload = {\"variants\": [spdi_notation]}\n",
    "\n",
    "    try:\n",
    "        response = requests.post(url, json=payload, headers=headers)\n",
    "        if response.status_code == 200:\n",
    "            return response.json()\n",
    "        else:\n",
    "            return {}\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred during the API call: {e}\")\n",
    "        return {}\n",
    "\n",
    "def write_to_csv(data, filename):\n",
    "    \"\"\"\n",
    "    Writes processed data to a CSV file.\n",
    "    \"\"\"\n",
    "    fieldnames = [\"chrom\", \"pos\", \"ref\", \"alt\", \"depth\", \"alt_reads\", \"percent_alt_reads\", \"percent_ref_reads\", \"gene\", \"variant_effect\", \"minor_allele\", \"minor_allele_frequency\", \"somatic\", \"id\"]\n",
    "    with open(filename, 'w', newline='') as csvfile:\n",
    "        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)\n",
    "        writer.writeheader()\n",
    "        for row in data:\n",
    "            writer.writerow(row)\n",
    "\n",
    "def process_ensembl_data(ensembl_data_item, vcf_data):\n",
    "    \"\"\"\n",
    "    Process a single item from Ensembl data to fit the CSV structure.\n",
    "\n",
    "    Args:\n",
    "        ensembl_data_item (dict): A single item from Ensembl response.\n",
    "        vcf_data (dict): The VCF data for the corresponding variant.\n",
    "\n",
    "    Returns:\n",
    "        dict: Processed Ensembl data for CSV output.\n",
    "    \"\"\"\n",
    "    processed_data = {\n",
    "        \"gene\": '',\n",
    "        \"variant_effect\": '',\n",
    "        \"minor_allele\": '',\n",
    "        \"minor_allele_frequency\": '',\n",
    "        \"somatic\": '',\n",
    "        \"alt_reads\": '',  # Placeholder for alternate reads\n",
    "        \"percent_alt_reads\": '',\n",
    "        \"percent_ref_reads\": ''\n",
    "    }\n",
    "\n",
    "    # Extracting gene and variant effect\n",
    "    if 'transcript_consequences' in ensembl_data_item:\n",
    "        first_consequence = ensembl_data_item['transcript_consequences'][0]\n",
    "        processed_data['gene'] = first_consequence.get('gene_symbol', '')\n",
    "        processed_data['variant_effect'] = first_consequence.get('impact', '')\n",
    "\n",
    "    # Extracting minor allele and its frequency\n",
    "    colocated_variants = ensembl_data_item.get('colocated_variants', [])\n",
    "    if colocated_variants:\n",
    "        first_variant = colocated_variants[0]\n",
    "        processed_data['minor_allele'] = first_variant.get('minor_allele', '')\n",
    "        processed_data['minor_allele_frequency'] = first_variant.get('minor_allele_freq', '')\n",
    "    \n",
    "    # Check if the variant is somatic\n",
    "    processed_data['somatic'] = '1' if any('somatic' in var for var in colocated_variants) else ''\n",
    "\n",
    "    # Calculating alternate and reference reads percentages\n",
    "    if vcf_data['depth']:\n",
    "        processed_data['alt_reads'] = vcf_data.get('alt_reads', '')\n",
    "        if processed_data['alt_reads']:\n",
    "            depth = int(vcf_data['depth'])\n",
    "            alt_reads = int(processed_data['alt_reads'])\n",
    "            processed_data['percent_alt_reads'] = str(round(alt_reads / depth * 100, 2))\n",
    "            processed_data['percent_ref_reads'] = str(round((depth - alt_reads) / depth * 100, 2))\n",
    "\n",
    "    # Calculate percent_alt_reads and percent_ref_reads\n",
    "    depth = int(vcf_data.get('depth', 0))\n",
    "    alt_reads = int(vcf_data.get('alt_reads', 0))\n",
    "    if depth > 0:\n",
    "        processed_data['percent_alt_reads'] = str(round(alt_reads / depth * 100, 2))\n",
    "        processed_data['percent_ref_reads'] = str(round((depth - alt_reads) / depth * 100, 2))\n",
    "\n",
    "    return processed_data\n",
    "\n",
    "\n",
    "def main(vcf_file, output_csv, output_pkl, full=False):\n",
    "    \"\"\"\n",
    "    Main function to process VCF file and write to CSV and pickle files.\n",
    "    \"\"\"\n",
    "    vcf_reader = cyvcf2.VCF(vcf_file)\n",
    "    all_data = []\n",
    "\n",
    "    count = 0\n",
    "    for record in vcf_reader:\n",
    "        \n",
    "        if full :\n",
    "            for vcf_data in parse_vcf_entry(record):\n",
    "                ensembl_data_list = query_ensembl(vcf_data)\n",
    "\n",
    "                if isinstance(ensembl_data_list, list) and ensembl_data_list:\n",
    "                    for ensembl_data in ensembl_data_list:\n",
    "                        combined_data = {**vcf_data, **process_ensembl_data(ensembl_data, vcf_data)}\n",
    "                        all_data.append(combined_data)\n",
    "                else:\n",
    "                    combined_data = {**vcf_data}\n",
    "                    all_data.append(combined_data)\n",
    "        else :\n",
    "            for vcf_data in parse_vcf_entry(record):\n",
    "                ensembl_data_list = query_ensembl(vcf_data)\n",
    "\n",
    "                if isinstance(ensembl_data_list, list) and ensembl_data_list:\n",
    "                    for ensembl_data in ensembl_data_list:\n",
    "                        combined_data = {**vcf_data, **process_ensembl_data(ensembl_data, vcf_data)}\n",
    "                        all_data.append(combined_data)\n",
    "                else:\n",
    "                    combined_data = {**vcf_data}\n",
    "                    all_data.append(combined_data)\n",
    "            count = count + 1\n",
    "            if count > 20:\n",
    "                break\n",
    "\n",
    "    with open(output_pkl, 'wb') as output_file:\n",
    "        pickle.dump(all_data, output_file)\n",
    "\n",
    "    write_to_csv(all_data, output_csv)\n",
    "\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "main(\"/qmd/VCF/test_vcf.txt\", \"/qmd/VCF/outputTest.csv\", \"/qmd/VCF/outputTest.pkl\", full=False)\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
